cupy/core/_reduction.pyx

```diff
@@ -284,21 +284,29 @@ cdef class _AbstractReductionKernel:
             internal.prod_sequence(out_shape),
             contiguous_size)
 
+        in_args_ = [Arg.from_obj(a) for a in in_args]
+
+        for i in range(self.nin):
+            arg = in_args_[i]
+            if arg.is_scalar():
+                (<ScalarArg>arg).apply_dtype(in_types[i])
+
         # Kernel arguments passed to the __global__ function.
-        inout_args = (
-            in_args
+        inout_args = tuple(
+            in_args_
             + out_args
             + [
-                _carray.Indexer(in_shape),
-                _carray.Indexer(out_shape),
+                IndexerArg(in_shape),
+                IndexerArg(out_shape),
                 # block_stride is passed as the last argument.
-                _scalar.CScalar.from_int32(block_stride),
+                # _scalar.CScalar.from_int32(block_stride),
+                Arg.from_obj(numpy.int32(block_stride)),
```

Comment:
Please note that this fix is just tentative, because it might have performance implication.
Perhaps you should either
* Make a measurement and confirm it's negligible (I think it's unlikely)
* Make a fast shortcut constructor of `Arg` that does not require `numpy.int32` (like `CScalar.from_int32`).
* Or perhaps treat this argument (and indexers) specially, instead of just the same way as array/scalar arguments. I think this option is desirable because it doesn't make sense to include their `get_immutable_key()`s in the cache key. But it would require a bit further design reconsideration.

Reply:
I left a TODO comment. I'll address your comment in another PR.

Reply:
OK. Please note that that means this PR introduces performance degradation.